packages/flutter/lib/src/gestures/mouse_cursor.dart

```diff
@@ -0,0 +1,332 @@
+  String describeCursor();
+
+  @override
+  String toString() {
+    return '$runtimeType(${describeCursor()})';
+  }
+}
+
+/// A mouse cursor that does nothing when activated.
+///
+/// See also:
+///
+///  * [SystemMouseCursors.releaseControl], which is an object of this cursor.
+@immutable
+class NoopMouseCursor extends MouseCursor {
+  /// Create a [NoopMouseCursor].
+  const NoopMouseCursor();
+
+  /// Does nothing and immediately returns true.
+  @override
+  Future<bool> activate(MouseCursorActivateDetails details) async {
+    return true;
+  }
+
+  @override
+  String describeCursor() => 'noop';
+}
+
+// A mouse cursor based on resources provided by the platform.
+//
+// See also:
+//
+//  * [SystemMouseCursors], which lists all system mouse cursors.
+@immutable
+class _SystemMouseCursor extends MouseCursor {
+  const _SystemMouseCursor(this.systemShape, this.description)
+    : assert(systemShape != null), assert(description != null);
+
+  final MouseCursorSystemShape systemShape;
+
+  final String description;
+
+  @override
+  Future<bool> activate(MouseCursorActivateDetails details) {
+    return details.platformDelegate.activateSystemCursor(
+      MouseCursorPlatformActivateSystemCursorDetails(
+        device: details.device,
+        systemShape: systemShape,
+      ),
+    );
+  }
+
+  @override
+  String describeCursor() {
+    return description;
+  }
+}
+
+// A [_SystemMouseCursor] that guarantees to be implemented.
+@immutable
+class _EnsuredImplementedSystemMouseCursor extends _SystemMouseCursor {
+  const _EnsuredImplementedSystemMouseCursor(
+    MouseCursorSystemShape systemShape,
+    String description,
+  ) : super(systemShape, description);
+
+  @override
+  Future<bool> activate(MouseCursorActivateDetails details) async {
+    final bool implemented = await super.activate(details);
+    assert(implemented);
+    return implemented;
+  }
+}
+
+/// The base class of a manager that maintains states related to mouse cursor
+/// and provides a simple interface to operate [MouseCursor]s.
+///
+/// Widgets should not use [MouseCursorManager] directly, instead they should
+/// assign [MouseCursor]s to regions, and then [MouseTracker] will handle cursor
+/// changes accordingly.
+///
+/// See also:
+///
+///  * [MouseCursor], which talks more about handling mouse cursors.
+///  * [StandardMouseCursorManager], which implements the platform-specific
+///    code based on the platform that this program is running on.
+///  * [MouseTracker], which uses this class.
+abstract class MouseCursorManager {
+  /// The delegate of the platform that this manager operates.
+  ///
+  /// It is provided to [MouseCursor] to perform platform operations.
+  MouseCursorPlatformDelegate get platformDelegate;
+
+  /// Set the cursor of pointer `device` to `cursor`.
+  ///
+  /// This method handles states or fallbacks.
+  ///
+  /// This method resolves if the operation is successful, or throws errors if
```

Comment:
It seems that the method will just switch to the basic cursor if an error occurs when setting the specified cursor?